packages/phpstan-rules/src/Rules/RequireStringArgumentInConstructorRule.php

```diff
@@ -0,0 +1,157 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Symplify\PHPStanRules\Rules;
+
+use PhpParser\Node;
+use PhpParser\Node\Arg;
+use PhpParser\Node\Expr\ClassConstFetch;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Scalar\String_;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\TypeWithClassName;
+use Symplify\Astral\Naming\SimpleNameResolver;
+use Symplify\PHPStanRules\Exception\ShouldNotHappenException;
+use Symplify\RuleDocGenerator\Contract\ConfigurableRuleInterface;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\ConfiguredCodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * Useful for prefixed phar bulid, to keep original references to class un-prefixed
+ *
+ * @see \Symplify\PHPStanRules\Tests\Rules\RequireStringArgumentInConstructorRule\RequireStringArgumentInConstructorRuleTest
+ */
+final class RequireStringArgumentInConstructorRule extends AbstractSymplifyRule implements ConfigurableRuleInterface
+{
+    /**
+     * @var string
+     */
+    public const ERROR_MESSAGE = 'Use quoted string in constructor "new %s()" argument on position %d instead of "::class. It prevent scoping of the class in building prefixed package.';
+
+    /**
+     * @var array<string, array<int>>
+     */
+    private $stringArgPositionsByType = [];
+
+    /**
+     * @var SimpleNameResolver
+     */
+    private $simpleNameResolver;
+
+    /**
+     * @param array<string, array<int>> $stringArgPositionsByType
+     */
+    public function __construct(SimpleNameResolver $simpleNameResolver, array $stringArgPositionsByType = [])
+    {
+        $this->stringArgPositionsByType = $stringArgPositionsByType;
+        $this->simpleNameResolver = $simpleNameResolver;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [New_::class];
+    }
+
+    /**
+     * @param New_ $node
+     * @return string[]
+     */
+    public function process(Node $node, Scope $scope): array
+    {
+        $errorMessages = [];
+
+        foreach ($this->stringArgPositionsByType as $type => $positions) {
+            if (! $this->isNodeVarType($node, $scope, $type)) {
```

Comment:
The `Symplify\PHPStanRules\TypeAnalyzer\ObjectTypeAnalyzer` can be used instead

Reply:
Done